### ISSUE TITLE
0.2.18 Matching URL problem

### ISSUE THREAD

**Issue by chenav** [angular-ui/ui-router#2549]
Hey,

Since 0.2.18, the URL matching does not catch up some URL 
In the following example, when LinkedIn API calls back a sequential URL, ui-router catches it and proceeds.
When Facebook API calls back a URL with a query string (2 arguments), it is never catched.

It works perfectly in 0.2.15 / 0.2.16 / 0.2.17 and the controller of the called view does the cookie context thing, it has to do.
My fallback state is called in 0.2.18, and the authentication is not completed then.

``` javascript
 state('callbackLinkedin', {
    url: '/callbackLK/:from/:token',
    templateUrl: function () {
        return 'partials/commons/pages/authentication/callback-linkedin.html';
    }
}).
state('callbackFacebook', {
    url: '/callbackFB?rp&code',
    templateUrl: function () {
        return 'partials/commons/pages/authentication/callback-facebook.html';
    }
}) 
```

I guess there is something new in the UrlMatcher 0.2.18 that does not handle the case the same way as before.
I rolled back to 0.2.17 to make my app work.

C.

**Comment by christopherthielen** [comment]
@chenav  can you show me an example facebook callback url?

**Comment by chenav** [comment]
The callbackURL that does not work is like so; 

```javascript
https://mydev.env.io/callbackFB?rp=account&code=AQDkJKAIxFhNAzaFPWchT5ybty7EVKgSJ-r57OR1R0isVhL1nJfi8HUlXGDw3h5-Oes0z1LLQAYzwH_YrcEUo5YXktQYbklDaObmiIWlVORytq_DzB4yZRDCxDr4WrH_9rh651dvcD0xYcTsNSLujDZxBbF51LfI-gR1ekS5hF-20MyBtMu6sMyl8LslvGG4skn-l1VnbZlwT3yPD5PFXxLJ-_b03nIlz-ZtOHtjr8Z57FulyZuW7NFTR_rXNvGAsswTpKkWC88OXNZ2D0Ycf0B8vCPRCiF053BTfQ5-YKQNoc1hCUAf-xET59K6l9o5FQ0#_=_
```

**Comment by christopherthielen** [comment]
So, this is OAUTH or something? Facebook redirects the browser to that url, your app loads from the server into the browser (html5 mode), then ui-router bootstraps and checks the URL, but the parameters don't match?

**Comment by chenav** [comment]
Yes, this is oauth and the URL is the callback URL.
And you are right about the sequence of calls. 
ui-router 0.2.18 does not match the URL. 
All the previous versions do match it.

**Comment by christopherthielen** [comment]
This seems to work for me.

http://plnkr.co/edit/wy1i9jSOOsmb2rLUpeN3?p=preview

**Comment by chenav** [comment]
Thank you for your feedback, Christopher.
Actually, it gave me headache because it was still not working on my side. 
I traced all my states and state params to see what was different with 0.2.18. 
And I finally found it.

I have a lot of states but only 2 with URL+query strings to match at router level. 
Here they are (note the order of the declarations) : 

```javascript
state('home', {
    url: '/:sitehash?appMode&refresh',
    templateUrl: function () {
            return 'template1.html';            
    }
}).
state('callbackFacebook', {
    url: '/callbackFB?rp&code',
    templateUrl: function () {
        return 'template2.html';
    }
})
```

In that order, it works with all version of ui-router inferior to 0.2.18
With 0.2.18, the Facebook Callback we talked about is matched as state 'home' with the following params:
```javascript
sitehash = 'callbackFB' 
appMode = undefined
refresh = undefined
rp = account
code = the_very_long_fbcode
```
It is a very tricky case and I am not sure it is what we should expect, as the signatures of the URLs (and the query strings) are different.

But in order to correct the behaviour, I just had to inverse the order of my 'home' state and my 'callbackFacebook' state.
With the 'callbackFacebook' state in first position, it matched it right with 0.2.18 

I do understand that one is matched dynamically when the other is not, but anyway, it was working right before that version. 
Now, since 0.2.18, it seems that we have to be extra careful with the order of states declaration .
Is that done on purpose ?
